benchmark/runner/mini_runners.cpp

```diff
@@ -0,0 +1,1161 @@
+  // constraint, but that caused unnecessary variation between clang and
+  // gcc benchmarks.
+  asm volatile("" ::"r"(datum));
+}
+
+/**
+ * To prevent the iterator op i from being optimized out on
+ * clang/gcc compilers, gcc requires noinline and noclone
+ * to prevent any inlining. noclone is used to prevent gcc
+ * from doing interprocedural constant propagation.
+ *
+ * DoNotOptimizeAway is placed inside the for () loop to
+ * ensure that the compiler does not blindly optimize away
+ * the for loop.
+ *
+ * TIGHT_LOOP_OPERATION(uint32_t, PLUS, +)
+ * Defines a function called uint32_t_PLUS that adds integers.
+ */
+#ifdef __clang__
+#define TIGHT_LOOP_OPERATION(type, name, op)                       \
+  __attribute__((noinline)) type __##type##_##name(size_t count) { \
+    type iterator = 1;                                             \
+    for (size_t i = 1; i <= count; i++) {                          \
+      iterator = iterator op i;                                    \
+      DoNotOptimizeAway(i);                                        \
+      DoNotOptimizeAway(iterator);                                 \
+    }                                                              \
+                                                                   \
+    return iterator;                                               \
+  }
+
+#elif __GNUC__
+#define TIGHT_LOOP_OPERATION(type, name, op)                                \
+  __attribute__((noinline, noclone)) type __##type##_##name(size_t count) { \
+    type iterator = 1;                                                      \
+    for (size_t i = 1; i <= count; i++) {                                   \
+      iterator = iterator op i;                                             \
+      DoNotOptimizeAway(i);                                                 \
+      DoNotOptimizeAway(iterator);                                          \
+    }                                                                       \
+                                                                            \
+    return iterator;                                                        \
+  }
+
+#endif
+
+/**
+ * Arg <0, 1, 2>
+ * 0 - Number of columns
+ * 1 - Number of rows
+ * 2 - Cardinality
+ */
+static void GenScanArguments(benchmark::internal::Benchmark *b) {
+  auto num_cols = {1, 4, 8, 15};
+  auto num_rows = {
+      std::pair<int, int>{10, 1},           std::pair<int, int>{10, 2},          std::pair<int, int>{10, 8},
+      std::pair<int, int>{100, 16},         std::pair<int, int>{100, 32},        std::pair<int, int>{100, 64},
+      std::pair<int, int>{10000, 128},      std::pair<int, int>{10000, 1024},    std::pair<int, int>{10000, 4096},
+      std::pair<int, int>{10000, 8192},     std::pair<int, int>{1000000, 1024},  std::pair<int, int>{1000000, 16384},
+      std::pair<int, int>{1000000, 131072}, std::pair<int, int>{1000000, 524288}};
```

Comment:
I think we can just generate an exponentially increased array for the cardinality based on the row num. Just similar to what I did [here](https://github.com/malin1993ml/terrier/blob/modeling-metrics/script/model/execution_engine_runner/generate_scan.py#L61). I think manually maintaining these pairs can be cumbersome if we want to boost the amount of training data later.

Reply:
Changed.